src/neurostatslib/basis.py

```diff
@@ -11,315 +11,902 @@
-            Spacing for basis functions, holding elements on the interval [0,
-            window_size). A good default is np.arange(window_size).
+        sample_pts : (number of samples, )
+            Spacing for basis functions, holding elements on the interval [min(sample_pts),
+            max(sample_pts)].
 
         Returns
         -------
-        basis_funcs : (n_basis_funcs, n_pts)
+        basis_funcs : (number of basis, number of samples)
             Evaluated spline basis functions.
 
         """
-        super().gen_basis_funcs(sample_pts)
-        sample_pts = sample_pts / self.window_size
+
+        # add knots if not passed
+        self._generate_knots(
+            sample_pts[0], perc_low=0.0, perc_high=1.0, is_cyclic=False
+        )
 
         return np.stack(
             [
-                mspline(sample_pts, self.order, i, self.knot_locs)
-                for i in range(self.n_basis_funcs)
+                mspline(sample_pts[0], self._order, i, self.knot_locs)
+                for i in range(self._n_basis_funcs)
             ],
             axis=0,
         )
 
+    def _check_n_basis_min(self) -> None:
+        """
+        Check that the user required enough basis elements. Most of the basis work with at least 1 element, but some
+        such as the RaisedCosineBasisLog requires a minimum of 2 basis to be well defined.
+
+        Raises
+        ------
+        ValueError
+            If an insufficient number of basis element is requested for the basis type
+        """
+        if self._n_basis_funcs < 1:
+            raise ValueError(
+                f"Object class {self.__class__.__name__} requires >= 1 basis elements. {self._n_basis_funcs} basis elements specified instead"
+            )
+
+
+class RaisedCosineBasis(Basis, abc.ABC):
+    def __init__(self, n_basis_funcs: int) -> None:
+        super().__init__(n_basis_funcs)
+        self._n_input_samples = 1
+
+    @abc.abstractmethod
+    def _transform_samples(self, sample_pts: NDArray) -> NDArray:
+        """
+        Abstract method for transforming sample points.
+
+        Parameters
+        ----------
+        sample_pts : (number of samples, )
+           The sample points to be transformed.
+        """
+        pass
+
+    def _evaluate(self, *sample_pts: NDArray) -> NDArray:
+        """Generate basis functions with given samples.
+
+        Parameters
+        ----------
+        sample_pts : (number of samples,)
+            Spacing for basis functions, holding elements on interval [0,
+            1). A good default is
+            ``nsl.sample_points.raised_cosine_log`` for log spacing (as used in
+            [2]_) or ``nsl.sample_points.raised_cosine_linear`` for linear
+            spacing.
+
+        Returns
+        -------
+        basis_funcs : (number of basis, number of samples)
+            Raised cosine basis functions
+
+        """
+
+        if any(sample_pts[0] < -1e-12) or any(sample_pts[0] > 1 + 1e-12):
```

Comment:
feels weird to hardcode the epsilon like this -- it should depend on the precision of the dtype right?